packages/test/src/test-typed-search-attributes.ts

```diff
@@ -0,0 +1,437 @@
+    inputTestCases.map(async ({ input, expected }) => {
+      const { taskQueue } = helpers(t);
+      const client = t.context.env.client;
+      const action: ScheduleOptionsAction = {
+        type: 'startWorkflow',
+        workflowType: dummyWorkflow,
+        taskQueue,
+      };
+      const handle = await client.schedule.create({
+        scheduleId: randomUUID(),
+        spec: {
+          calendars: [{ hour: { start: 2, end: 7, step: 1 } }],
+        },
+        action,
+        ...input,
+      });
+      const desc = await handle.describe();
+      t.deepEqual(desc.searchAttributes, expected.searchAttributes);
+      t.deepEqual(desc.typedSearchAttributes, expected.typedSearchAttributes);
+    })
+  );
+});
+
+export const getWorkflowInfo = defineQuery<WorkflowInfo>('getWorkflowInfo');
+export const mutateSearchAttributes =
+  defineSignal<[SearchAttributes | SearchAttributeUpdatePair[]]>('mutateSearchAttributes');
+export const complete = defineSignal('complete');
+
+export async function changeSearchAttributes(): Promise<void> {
+  let isComplete = false;
+  setHandler(getWorkflowInfo, () => {
+    return workflowInfo();
+  });
+  setHandler(complete, () => {
+    isComplete = true;
+  });
+  setHandler(mutateSearchAttributes, (attrs) => {
+    upsertSearchAttributes(attrs);
+  });
+  await condition(() => isComplete);
+}
+
+test('upsert works with various search attribute mutations', async (t) => {
+  const { createWorker, startWorkflow } = helpers(t);
+  const worker = await createWorker({ namespace: t.context.env.namespace });
+  await worker.runUntil(async () => {
+    // Start workflow with some initial search attributes.
+    const handle = await startWorkflow(changeSearchAttributes, {
+      typedSearchAttributes: typedAttrsListInput,
+    });
+    let res = await handle.query(getWorkflowInfo);
+    let desc = await handle.describe();
+    assertWorkflowInfoSearchAttributes(t, res, untypedFromTypedInput, typedAttrsListInput);
+    assertWorkflowDescSearchAttributes(t, desc, untypedFromTypedInput, typedAttrsListInput);
+
+    // Update search attributes with untyped input.
+    const untypedUpdateAttrs: SearchAttributes = {
+      typed_text: ['new_value'],
+      typed_keyword: ['new_keyword'],
+      typed_int: [2],
+      typed_double: [2.34],
+      typed_datetime: [secondDate],
+      typed_keyword_list: ['three', 'four', 'five'],
+      // Delete key - empty value.
+      typed_bool: [],
+    };
+
+    // Update search attributes with untyped input.
+    await handle.signal(mutateSearchAttributes, untypedUpdateAttrs);
+    res = await handle.query(getWorkflowInfo);
+    desc = await handle.describe();
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const { typed_bool, ...untypedUpdateExpected } = untypedUpdateAttrs;
+
+    assertWorkflowInfoSearchAttributes(t, res, untypedUpdateExpected, [
+      { key: { name: 'typed_text', type: SearchAttributeType.TEXT }, value: 'new_value' },
```

Comment:
Make sure to use `defineSearchAttribute()` in at least one test (though I would personally use it in _most_ tests).

Reply:
Changed the tests in `test-typed-search-attributes.ts` to use the helper.